plugins/woocommerce-blocks/assets/js/blocks/product-template/utils.tsx

```diff
@@ -0,0 +1,87 @@
+/**
+ * External dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useMemo } from '@wordpress/element';
+
+/**
+ * In Product Collection block, queryContextIncludes attribute contains
+ * list of attribute names that should be included in the query context.
+ *
+ * This hook returns the query context object based on the attribute names
+ * provided in the queryContextIncludes array.
+ *
+ * Example:
+ * {
+ * 	clientID = 'd2c7e34f-70d6-417c-b582-f554a3a575f3',
+ * 	queryContextIncludes = [ 'collection' ]
+ * }
+ *
+ * The hook will return the following query context object:
+ * {
+ *  collection: 'woocommerce/product-collection/featured'
+ * }
+ *
+ * @param args                      Arguments for the hook.
+ * @param args.clientId             Client ID of the inner block.
+ * @param args.queryContextIncludes Array of attribute names to be included in the query context.
+ *
+ * @return Query context object.
+ */
+export const useProductCollectionQueryContext = ( {
+	clientId,
+	queryContextIncludes,
+}: {
+	clientId: string;
+	queryContextIncludes: string[];
+} ) => {
+	const productCollectionBlockAttributes = useSelect(
+		( select ) => {
+			const { getBlockParentsByBlockName, getBlockAttributes } =
+				select( 'core/block-editor' );
+
+			const parentBlocksClientIds = getBlockParentsByBlockName(
+				clientId,
+				'woocommerce/product-collection',
+				true
+			);
+
+			if ( parentBlocksClientIds?.length ) {
+				const closestParentClientId = parentBlocksClientIds[ 0 ];
+				return getBlockAttributes( closestParentClientId );
+			}
+
+			return null;
+		},
+		[ clientId ]
+	);
+
+	return useMemo( () => {
+		// If the product collection block is not found, return null.
+		if ( ! productCollectionBlockAttributes ) {
+			return null;
+		}
+
+		/**
+		 * Initialize the query context object with collection and id attributes as
+		 * they should always be included in the query context.
+		 */
+		const queryContext: {
+			[ key: string ]: unknown;
+		} = {
+			collection: productCollectionBlockAttributes?.collection,
+			id: productCollectionBlockAttributes?.id,
+		};
```

Comment:
I'm not convinced these attributes should be handled in the hook. I'd lift that part upstream to `plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx` so the hook is generic and doesn't care/is not aware of some specific attributes. But I don't have a strong opinion here so feel free to ignore this comment.

Alternatively, I'd extract this mapping to some variable because I'm pretty sure we'll find more attributes like these that we will like to always include in query. In that case it would be easier to extend some standalone object  then finding this initialization here and extend it.

Reply:
Great suggestion! I like the idea of using a constant variable to keep track of the default query context attributes. I've implemented the changes in https://github.com/woocommerce/woocommerce/pull/44150/commits/d4f071d86319fb32e4d7d4aaa37620034bf6088d. Could you please have a quick look? 👀